test/bazel/proto/BUILD.bazel

```diff
@@ -17,3 +17,8 @@ typescript_proto_library(
     name = "pizza_service_ts_proto",
     proto = ":pizza_service_proto",
 )
+
+typescript_proto_library(
+    name = "external_ts_proto",
+    proto = "@external_protos//:external_proto",
```

Comment:
Can you add a similar ts proto lib for a google protobuf and use both of these new libs in a karma web test? [example](https://github.com/improbable-eng/ts-protoc-gen/blob/master/test/bazel/BUILD.bazel#L55)

Reply:
@Dig-Doug - I see you importing from an ambient module, but I don't see how/where those are defined. I tried using what I saw in the *_amd.js file, but that didn't seem to work. For instance, writing a test for this, I used `ts_protoc_gen/external_pb` but would get the error `Cannot find module 'ts_protoc_gen/external_pb'`

Reply:
So I managed to get it compiling by using a relative module. I think we still want to eventually switch it back though.

That said, when I run it as I have it now, I get this (rather unscrutable) error:

```
[33m29 04 2019 18:53:55.784:WARN [web-server]: [39m404: /ts_protoc_gen/external/external_protos/external_pb.js
HeadlessChrome 69.0.3497 (Mac OS X 10.14.4) ERROR: 'There is no timestamp for ts_protoc_gen/external/external_protos/external_pb.js!'

HeadlessChrome 69.0.3497 (Mac OS X 10.14.4) ERROR
  {
    "message": "Uncaught Error: Script error for \"ts_protoc_gen/external/external_protos/external_pb\", needed by: ts_protoc_gen/test/bazel/external_proto_with_deps_test.spec\nhttp://requirejs.org/docs/errors.html#scripterror\nat npm/node_modules/requirejs/require.js:143:9\n\nError: Script error for \"ts_protoc_gen/external/external_protos/external_pb\", needed by: ts_protoc_gen/test/bazel/external_proto_with_deps_test.spec\nhttp://requirejs.org/docs/errors.html#scripterror\n    at makeError (node_modules/requirejs/require.js:168:17)\n    at HTMLScriptElement.onScriptError (node_modules/requirejs/require.js:1738:36)",
    "str": "Uncaught Error: Script error for \"ts_protoc_gen/external/external_protos/external_pb\", needed by: ts_protoc_gen/test/bazel/external_proto_with_deps_test.spec\nhttp://requirejs.org/docs/errors.html#scripterror\nat npm/node_modules/requirejs/require.js:143:9\n\nError: Script error for \"ts_protoc_gen/external/external_protos/external_pb\", needed by: ts_protoc_gen/test/bazel/external_proto_with_deps_test.spec\nhttp://requirejs.org/docs/errors.html#scripterror\n    at makeError (node_modules/requirejs/require.js:168:17)\n    at HTMLScriptElement.onScriptError (node_modules/requirejs/require.js:1738:36)"
  }
```

Here are my bazel targets:

```Python
ts_library(
    name = "external_proto_with_deps_test",
    testonly = 1,
    srcs = ["external_proto_with_deps_test.spec.ts"],
    tsconfig = ":tsconfig.json",
    deps = [
        "//test/bazel/proto:external_ts_proto",
        "@npm//@types/jasmine",
        # Don't put external_dep_ts_proto here, we want to test it is included transitively
    ],
)

ts_web_test_suite(
    name = "external_proto_with_deps_test_suite",
    browsers = [
        "@io_bazel_rules_webtesting//browsers:chromium-local",
    ],
    deps = [
        ":google_protobuf_requirejs",
        ":external_proto_with_deps_test",
    ],
)
```

Reply:
So I'm wondering if the issue is that `external_pb_amd.js` specifies `define("ts_protoc_gen/external_pb", ...` for the module but the error output indicates it expects `ts_protoc_gen/external/external_protos/external_pb`. Indeed, when I use that as the import, it does compile, so that seems to be the registered ambient module, but I'm still getting errors at runtime.

Reply:
Yes, it looks like there is a problem with the external AMD modules names. I think it should be something like `external_protos/external_pb`. The format is `<workspace-name>/path/to/proto`, e.g. `ts_protoc_gen/test/bazel/proto/pizza_service_pb`.

Reply:
Okay, so that sounds like I should probably factor out some of this logic and change the arguments we're calling `ctx.executable._convert_to_amd` with.

Will that solve the problem completely though? If it builds, it's clearly finding the module, even if it's got the wrong name. What are the runtime errors I'm getting?

Reply:
This is starting to seem crazy, too. We can't be the only aspect that has this problem. It feels like stuff like `isExternal` and `workspace_name` should be available per dependency and not only from the original context. Is this a core bazel problem?

Reply:
Actually I was just thinking about this a bit more, one more different thought:

In the proto files, you don't specify the workspace when you import them. If you look at the generate js code from protoc, it also isn't including the workspace when looking for the modules. So what I'm thinking is that the AMD module names shouldn't include the workspace. (The file locations are fine I think, it just the names inside the files that need to change). What do you think?

Not sure if this will fix everything, haven't tried it locally.

Reply:
Actually I was just thinking about this a bit more, one more different thought:

In the proto files, you don't specify the workspace when you import them. If you look at the generate js code from protoc, it also isn't including the workspace when looking for the modules. So what I'm thinking is that the AMD module names shouldn't include the workspace. (The file locations are fine I think, it just the names inside the files that need to change). What do you think?

Not sure if this will fix everything, haven't tried it locally.

Reply:
I guess what happens if I have two dependencies from two different external workspaces that have the same name/location? There'd be no way to uniquely identify them. This isn't an issue I'm having (only importing from a single workspace atm) but it feels like workspace has to be necessary.

Reply:
Omg, this stuff is there and we're doing it in `proto_path`. Okay, that gives me a lot to go off of. I think I might be able to clean this up.

Reply:
I got it so that if you look at `external_pb_amd.js` the module is defined as `external/external_protos/external_pb` but I still get a compile error if I import that. The error output indicates it's the same as above: `ts_protoc_gen/external/external_protos/external_dep_pb`. Clearly, it's getting it from somewhere other than the _amd.js file, so any ideas on where that would be? I'm assuming it's based on the directory structure, but since it's compile time, the only way to get an ambient module is to import it via node, right?

Like I said earlier, I'm back to the part of the problem where it compiles and it still fails.

Reply:
I did some debugging, here's what I've found:

- Currently, I believe the only way to import code from an external repo in TS is to use a path like `workspace/external/external_repo/path`, e.g. `ts_protoc_gen/external/external_protos`. This looks like a bug to me in rules_typescript, (I think just `external_protos/` should be valid), but I'm not sure -- I'll create a reproduction and file an issue with them.

- When the typescript code is translated into JS, it uses the import path as the amd module name.  So, the AMD module name of the proto must be `ts_protoc_gen/external/external_protos/external_pb`.

For now, I think we should just make the proto AMD module name have the workspace name, `ts_protoc_gen/`, at the beginning. If in the future the external repo path is fixed, we can change the code to export AMD modules like `external_protos/external_pb`. What do you think?